highlight/src/lib.rs

```diff
@@ -109,7 +112,28 @@ struct HighlightIterLayer<'a, 'tree: 'a> {
     depth: usize,
 }
 
-impl Highlighter {
+// This enum needed to move flag ownership into HighlighIter struct instances
+// and to avoid an issue in Highlighter::highlight() that a local reference to
+// Arc<AtomicUsize> can't leave long enough to be used inside of HighlighIter::next()
+enum CancellationVariant<'c> {
+    Value(Arc<AtomicUsize>),
+    Borrow(&'c AtomicUsize),
+}
+
+impl<'c> CancellationVariant<'c> {
+    fn check(&self) -> Option<Error> {
+        let flag = match self {
+            CancellationVariant::Borrow(f) => f.load(Ordering::Relaxed),
+            CancellationVariant::Value(f) => f.as_ref().load(Ordering::Relaxed),
+        };
+        if flag != 0 {
+            return Some(Error::Cancelled);
+        }
+        None
+    }
+}
+
+impl<'a> Highlighter {
```

Comment:
I don't understand why we need to change the `Highlighter` API at all. I think it was already safe, which is why it wasn't marked as `unsafe`. The `Highlighter` encapsulates the `Parser`, so it can dynamically make sure that the `Parser` does not hold onto the cancellation flag longer than the reference is valid.

AFAICT, this API can stay the same as before, but just switched to using `set_cancellation_flag_unchecked` internally. Am I missing something?